samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote.couchbase;
+
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.error.TemporaryFailureException;
+import com.couchbase.client.java.error.TemporaryLockFailureException;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.List;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.context.Context;
+import org.apache.samza.operators.functions.ClosableFunction;
+import org.apache.samza.operators.functions.InitableFunction;
+import org.apache.samza.serializers.Serde;
+
+
+/**
+ * Base class for {@link CouchbaseTableReadFunction} and {@link CouchbaseTableWriteFunction}
+ * @param <V> Type of values to read from / write to couchbase
+ */
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
+  protected Duration timeout = Duration.ZERO; // default value 0 means no timeout
+  protected Duration ttl = Duration.ZERO; // default value 0 means no ttl, data will be stored forever
```

Comment:
These defaults are a bit counter-intuitive. 
You can use java's Optional<Duration> on timeout and ttl to mean that these features are enabled, else disabled.

Alternately, would it be possible for a negative value of timeout to mean "no timeout", similarly for ttl.
Duration does allow negative.

Reply:
These defaults are consistent with Couchbase client. Actually, even if I don't set ttl, it will still be 0 in a Couchbase document. So I use those default values to be consistent with Couchbase client API. Otherwise, we need to add extra logic to check those values in read/write function, making it a little more complicated. How do you think?

Reply:
I see, in that case, if you can please add the 0-special-case behavior to the javadoc for methods withTimeout() and withTtl().

Reply:
No problem.